chainermn/optimizers.py

```diff
@@ -20,14 +20,26 @@ def update(self, lossfun=None, *args, **kwds):
             loss.backward()
             del loss
 
-        if self.needs_broadcast:
+        if self.is_changed(target):
             self.communicator.broadcast_data(target)
-            super(_MultiNodeOptimizer, self).__setattr__(
-                'needs_broadcast', False)
         else:
             self.communicator.allreduce_grad(target)
             self.actual_optimizer.update(None, *args, **kwds)
 
+    def is_changed(self, target):
+        previous_params = self.target_params
+        super(_MultiNodeOptimizer, self).__setattr__(
+            'target_params', [(name, param.data is not None) for name, param in sorted(target.namedparams())])
+        if len(previous_params) == len(self.target_params):
+            for param1, param2 in zip(self.target_params, previous_params):
+                if param1[0] != param2[0]:
+                    return True
+                if param1[1] != param2[1]:
+                    return True
+            return False
+        else:
+            return True
```

Comment:
Let us make the case analysis as easy as possible. Here, how about using "early return" as follows (for details, please refer to the book 'The Art of Readable Code'):

```python
if len(previous_params) != len(self.target_params):
  return True

for param1, param2 in zip(self.target_params, previous_params):
  ...
```

Reply:
Thanks, I will fix it.